garrysmod/lua/includes/extensions/util/promise.lua

```diff
@@ -0,0 +1,193 @@
+    end, function(err)
+      if isfunction(onreject) then
+        local safe, args = pcall(function()
+          resolve(onreject(err))
+        end)
+        if not safe then reject(args) end
+      else reject(err) end
+    end)
+  end)
+end
+function Promise:Catch(onreject)
+  return self:Then(nil, onreject)
+end
+function Promise:Await()
+  if not coroutine.running() then return end
+  while self:IsPending() do coroutine.yield() end
+  if self:IsResolved() then
+    return self._value
+  else error(self._value) end
+end
+function Promise:__tostring()
+  if self:IsPending() then return "Promise: pending"
+  elseif self:IsResolved() then return "Promise: fulfilled ("..tostring(self._value)..")"
+  else return "Promise: rejected ("..tostring(self._value)..")" end
+end
+setmetatable(Promise, {__call = Promise._New})
+
+--[[---------------------------------------------------------
+    Returns a new Promise
+-----------------------------------------------------------]]
+function util.Promise(callback)
+  return Promise(callback)
+end
+
+--[[---------------------------------------------------------
+    Returns a Promise that is resolved when every Promise
+    passed as a parameter is resolved, or rejected
+    when one of the passed promises is rejected
+-----------------------------------------------------------]]
+function util.PromiseAll(promises)
+  return Promise(function(resolve, reject)
+    local remaining = table.Count(promises)
+    if remaining > 0 then
+      local results = {}
+      for key, promise in pairs(promises) do
+        promise:Then(function(res)
+          results[key] = res
+          remaining = remaining - 1
+          if remaining == 0 then
+            resolve(results)
+          end
+        end, reject)
+      end
+    else resolve({}) end
+  end)
+end
+
+--[[---------------------------------------------------------
+    Returns a Promise that is resolved/rejected
+    with the first settled Promise
+-----------------------------------------------------------]]
+function util.PromiseFirst(promises)
+  return Promise(function(resolve, reject)
+    for key, promise in pairs(promises) do
+      promise:Then(resolve, reject)
+    end
+  end)
+end
+
+--[[---------------------------------------------------------
+    Returns a Promise that is resolved when the function
+    passed as a parameter stops running, or rejected if an
+    error occurs during the coroutine
+-----------------------------------------------------------]]
+
+local id = 0
+local coroutines = {}
+
+function util.PromiseAsync(callback)
+  return Promise(function(resolve, reject)
+    local co = coroutine.create(function()
+      local safe, args = pcall(function()
+        resolve(callback())
+      end)
+      if not safe then reject(args) end
+    end)
+    coroutines[id] = co
+    id = id+1
+  end)
+end
+
+hook.Add("Think", "PromiseAsyncCoroutines", function()
+  for id, co in pairs(coroutines) do
+    if co == nil then continue end
```

Comment:
co will never be nil here - table values cannot be nil.

Reply:
I put that in here just in case, I'll remove it.